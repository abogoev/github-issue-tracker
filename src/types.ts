import { NativeStackScreenProps } from "@react-navigation/native-stack";

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export enum CurrentRoute {
  Home = "Home",
  Details = "Details",
}

export type StackParams = {
  [CurrentRoute.Home]: undefined;
  [CurrentRoute.Details]: { owner: string; repo: string; number: number };
};

export type HomeNavigationProps = NativeStackScreenProps<
  StackParams,
  typeof CurrentRoute.Home
>;

export type DetailsNavigationProps = NativeStackScreenProps<
  StackParams,
  typeof CurrentRoute.Details
>;

export type Issue = {
  number: number;
  title: string;
  user: {
    avatar_url: string;
  };
  body: string;
};
