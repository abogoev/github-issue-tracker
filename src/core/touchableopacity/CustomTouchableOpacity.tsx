import { TouchableOpacity, TouchableOpacityProps } from "react-native";
import React, { FC } from "react";
import { ACTIVE_OPACITY } from "../../utils/constants";

const CustomTouchableOpacity: FC<TouchableOpacityProps> = (props) => {
  return <TouchableOpacity activeOpacity={ACTIVE_OPACITY} {...props} />;
};

export default CustomTouchableOpacity;
