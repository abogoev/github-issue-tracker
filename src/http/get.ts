import { Issue } from "../types";
import axiosInstance from "./axiosInstance";

export const getIssuesByOwnerAndRepo = async (
  owner: string,
  repo: string,
  page: number
) => {
  const res = await axiosInstance.get<Issue[]>(
    `/repos/${owner}/${repo}/issues`,
    { params: { per_page: 50, page } }
  );
  console.log(res.data.length);
  return res.data;
};
