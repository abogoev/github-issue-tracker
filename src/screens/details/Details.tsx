import React, { useState, useEffect } from "react";
import {
  ScrollView,
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
} from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";
import Comment from "./infobox/Comment";
import theme from "../../theme/theme";
import { DetailsNavigationProps } from "../../types";
import { VFC } from "react";
import OpenClosedBadge from "./badge/OpenClosedBadge";
import BackIcon from "../../../assets/svg/back.svg";
import { fetchIssuesByOwnerAndRepo } from "../../http/get";

interface Props extends DetailsNavigationProps {
  a: boolean;
}

const Details: VFC<Props> = ({ route, navigation }) => {
  // console.log(route.params);
  const { number, owner, repo } = route.params;
  const [issue, setIssue] = useState(null);

  useEffect(() => {
    (async () => {
      try {
        await fetchIssuesByOwnerAndRepo("octocat", "hello-world");
      } catch (error) {
        console.log(error);
      }
    })();
  }, []);

  return (
    <SafeAreaView>
      <TouchableOpacity
        accessibilityRole="button"
        onPress={() => navigation.goBack()}
        style={styles.back}
      >
        <BackIcon width={32} height={32} />
      </TouchableOpacity>
      <ScrollView style={styles.container}>
        <Text style={styles.title}>About</Text>
        <OpenClosedBadge />
        <View style={styles.hr} />
        <Comment />
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  back: {
    alignSelf: "flex-start",
    marginLeft: 14,
    marginBottom: 16,
  },
  container: {
    paddingHorizontal: 20,
  },
  title: {
    ...theme.typography.verylargeText,
    color: theme.palette.veryDarkGrey,
    marginBottom: 16,
  },
  hr: {
    height: 1,
    backgroundColor: theme.palette.neutral,
    marginVertical: 16,
  },
});

export default Details;
