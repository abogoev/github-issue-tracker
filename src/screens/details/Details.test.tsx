import React from "react";
import "react-native";
import { act, create, ReactTestRenderer } from "react-test-renderer";
import Details from "./Details";
import { getIssue } from "../../http/get";
import { Issue } from "../../types";
import CustomTouchableOpacity from "../../core/touchableopacity/CustomTouchableOpacity";

jest.mock("@react-navigation/core");

jest.mock("../../http/get");

const mockedGetIssue = getIssue as jest.MockedFunction<typeof getIssue>;

describe("Details", () => {
  it("should match snapshot and call back correctly", async () => {
    const goBackMock = jest.fn();
    const props: any = {
      navigation: {
        goBack: goBackMock,
      },
      route: {
        params: { owner: "owner", repo: "repo", number: 1 },
      },
    };
    const data: Issue = {
      number: 1,
      title: "title",
      user: {
        login: "login",
        avatar_url: "",
      },
      body: "body",
      created_at: "2012-04-23T18:25:43.511Z",
      closed_at: null,
    };

    mockedGetIssue.mockResolvedValueOnce(data);

    let component: ReactTestRenderer | undefined;
    await act(async () => {
      component = await create(<Details {...props} />);
    });

    expect(goBackMock).not.toHaveBeenCalled();
    component?.root.findByType(CustomTouchableOpacity).props.onPress();
    expect(goBackMock).toHaveBeenCalledTimes(1);

    expect(component).toMatchSnapshot();
  });
});
