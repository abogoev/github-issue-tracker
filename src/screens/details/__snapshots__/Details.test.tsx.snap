// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details should match snapshot 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityRole="button"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "flex-start",
        "marginBottom": 16,
        "marginLeft": 14,
        "opacity": 1,
      }
    }
  >
    <svg
      data-file-name="SvgBack"
      height={32}
      width={32}
    />
  </View>
  <RCTScrollView
    style={
      Object {
        "paddingHorizontal": 20,
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "#24292f",
            "fontFamily": "Helvetica",
            "fontSize": 24,
            "fontWeight": "300",
            "lineHeight": 36,
            "marginBottom": 16,
          }
        }
      >
        title
      </Text>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "flex-start",
              "backgroundColor": "rgb(45, 164, 78)",
              "borderRadius": 28,
              "flexDirection": "row",
              "height": 32,
              "paddingHorizontal": 12,
            },
            undefined,
          ]
        }
      >
        <svg
          data-file-name="SvgOpenIssueWhite"
        />
        <Text
          style={
            Object {
              "color": "#fff",
              "fontFamily": "Helvetica",
              "fontSize": 14,
              "fontWeight": "500",
              "lineHeight": 20,
            }
          }
        >
           Open
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "rgb(208, 215, 222)",
            "height": 1,
            "marginVertical": 16,
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            undefined,
          ]
        }
      >
        <Image
          accessibilityIgnoresInvertColors={true}
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "borderRadius": 20,
              "height": 40,
              "marginRight": 16,
              "width": 40,
            }
          }
        />
        <View
          style={
            Object {
              "borderColor": "rgb(208, 215, 222)",
              "borderRadius": 6,
              "borderWidth": 1,
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(246, 248, 250)",
                "borderBottomWidth": 1,
                "borderColor": "rgb(208, 215, 222)",
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
                "minHeight": 39,
                "paddingHorizontal": 16,
                "paddingVertical": 8,
              }
            }
          >
            <Text>
              login
            </Text>
            <Text>
              commented on 
              23/04/2012, 21:25:43
            </Text>
          </View>
          <View
            style={
              Object {
                "minHeight": 53,
                "padding": 16,
              }
            }
          >
            <Text>
              body
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;

exports[`Details should match snapshot and call back correctly 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityRole="button"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "flex-start",
        "marginBottom": 16,
        "marginLeft": 14,
        "opacity": 1,
      }
    }
  >
    <svg
      data-file-name="SvgBack"
      height={32}
      width={32}
    />
  </View>
  <RCTScrollView
    style={
      Object {
        "paddingHorizontal": 20,
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "#24292f",
            "fontFamily": "Helvetica",
            "fontSize": 24,
            "fontWeight": "300",
            "lineHeight": 36,
            "marginBottom": 16,
          }
        }
      >
        title
      </Text>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "flex-start",
              "backgroundColor": "rgb(45, 164, 78)",
              "borderRadius": 28,
              "flexDirection": "row",
              "height": 32,
              "paddingHorizontal": 12,
            },
            undefined,
          ]
        }
      >
        <svg
          data-file-name="SvgOpenIssueWhite"
        />
        <Text
          style={
            Object {
              "color": "#fff",
              "fontFamily": "Helvetica",
              "fontSize": 14,
              "fontWeight": "500",
              "lineHeight": 20,
            }
          }
        >
           Open
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "rgb(208, 215, 222)",
            "height": 1,
            "marginVertical": 16,
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            undefined,
          ]
        }
      >
        <Image
          accessibilityIgnoresInvertColors={true}
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "borderRadius": 20,
              "height": 40,
              "marginRight": 16,
              "width": 40,
            }
          }
        />
        <View
          style={
            Object {
              "borderColor": "rgb(208, 215, 222)",
              "borderRadius": 6,
              "borderWidth": 1,
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(246, 248, 250)",
                "borderBottomWidth": 1,
                "borderColor": "rgb(208, 215, 222)",
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
                "minHeight": 39,
                "paddingHorizontal": 16,
                "paddingVertical": 8,
              }
            }
          >
            <Text>
              login
            </Text>
            <Text>
              commented on 
              23/04/2012, 21:25:43
            </Text>
          </View>
          <View
            style={
              Object {
                "minHeight": 53,
                "padding": 16,
              }
            }
          >
            <Text>
              body
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
